styx-scheduler-service/src/main/java/com/spotify/styx/Scheduler.java

```diff
@@ -205,6 +216,64 @@ private void evaluateResourcesForDequeue(
         .map(resource -> ResourceWithInstance.create(resource, instanceState));
   }
 
+  @VisibleForTesting
+  void triggerBackfills(Collection<InstanceState> activeStates) {
+    final List<Backfill> backfills;
+    try {
+      backfills = storage.backfills();
+    } catch (IOException e) {
+      LOG.warn("Failed to get resource limits", e);
+      return;
+    }
+
+    final Map<String, Long> backfillStates = activeStates.stream()
+        // TODO: find out if backfill in other way that string comparison
+        .filter(state -> state.runState().data().triggerId().orElse("").startsWith("backfill-"))
+        .collect(groupingBy(
+            state -> state.runState().data().triggerId().orElse("UNKNOWN"),
+            HashMap::new,
+            counting()));
+
+    backfills.forEach(backfill -> {
+      final Optional<Workflow> workflowOpt = workflowCache.workflow(backfill.workflowId());
+
+      if (!workflowOpt.isPresent()) {
+        LOG.warn("workflow not found for backfill: {}", backfill);
+        return;
+      }
+
+      final Workflow workflow = workflowOpt.get();
+      final int needed = backfill.concurrency() - backfillStates.getOrDefault(backfill.id(), 0L).intValue();
+      if (needed <= 0) {
+        return;
+      }
+
+      final List<Instant> parameters =
+          ParameterUtil.rangeOfInstants(backfill.nextTrigger(), backfill.end(),
+                                        workflow.schedule().partitioning())
+              .stream()
+              .limit(needed + 1)
+              .collect(toList());
+
+      if (parameters.isEmpty()) {
+        return;
+      }
+
+      final List<Instant> triggers = parameters.subList(0, Math.min(needed, parameters.size()));
+
+      triggers.forEach(
+          partition -> triggerListener.event(workflow, Trigger.backfill(backfill.id()), partition));
```

Comment:
This will throw runtime `AlreadyInitializedException` if already active, we should probably catch that, logging the event internally and proceed. `--force-halt` is out of scope for this PR, right?